### PR TITLE
feat(search-backend): Add spec and update to use ApiRouter typing.

### PR DESCRIPTION
**Pull request by sennyeya**
## Hey, I just made a Pull Request!

- Added an OpenAPI 3.0 spec and typed the router with the `ApiRouter` type. 
  - Creating as 3.0 instead of 3.1 due to lack of support for tooling (see #17482) to use 3.1.
- Fixed a bug in the type resolution of query parameters, where parameters that had internal schema references weren't resolving correctly.

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [ ] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Comment by freben** [comment]
See https://github.com/backstage/backstage/pull/17557

**Comment by freben** [comment]
(also the e2e tests are fixed locally now, feel free to rebase on fresh master, especially after the above PR is merged)

**Comment by sennyeya** [comment]
@freben Updated with the new `.generated` format and rebased for the E2E tests.

File: plugins/search-backend/src/schema/openapi.yaml
Comment (freben) [inline plugins/search-backend/src/schema/openapi.yaml:26]: I just want to say that this in general isn't true for JsonObject - anything json serializable can go as values including arrays and null. So leaving this entirely unspecified/open should be OK right?

**Review by freben — APPROVED** [review]
Looks good I think

**Comment by jhaals** [comment]
@backstage/discoverability-maintainers please take a look and feel free to 🚢 if this looks good with you!

File: plugins/search-backend/src/schema/openapi.yaml
Comment (sennyeya) [reply plugins/search-backend/src/schema/openapi.yaml:26]: Thanks for pointing that out, I'd meant to call that out initially. The existing zod schema was resolving recursively n times, which isn't possible to resolve with the current json-schema-to-ts library we're using. It also causes a fairly significant performance issue when resolving that recursive type even if we can't resolve it. I toyed around with using a 2 or 3 level deep resolution, but I think it makes more sense to just leave the schema fairly open here and keep the zod enforcement as is. Will add a comment here to encapsulate this.

Comment (freben) [reply plugins/search-backend/src/schema/openapi.yaml:26]: Yeah is it possible to even do `additionalProperties: {}`? I mean this data is by its very nature going to be JSON serializable anyway.

**Review by aitherios — APPROVED** [review]

**Comment by aitherios** [comment]
🚢

**Pull request merged**
